### PR TITLE
add-roles-solution

### PR DESCRIPTION
**Pull request by Valdemar3X**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/api/auth/**", "/error")
+                                .permitAll()
+                                .requestMatchers("/swagger-ui/**")
+                                .permitAll()
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]: why do we need permitAll() 2 times?

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -1,29 +1,46 @@
...
 import mate.academy.bookstore.model.User;
+import mate.academy.bookstore.repository.role.RoleRepository;
 import mate.academy.bookstore.repository.user.UserRepository;
 import mate.academy.bookstore.service.UserService;
-import org.springframework.stereotype.Service;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
 
-@Service
+@Component
```
Comment (Rommelua) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]: ??
Service is better

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_user_id
+                    references: users(id)
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_role_id
+                    references: roles(id)
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: these two columns combined need to be a primary key for this table

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -1,29 +1,46 @@
...
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
-    private final UserRepository repository;
+    private final UserRepository userRepository;
     private final UserMapper mapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto request)
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java:27]: @Transactional. You make more than one call to DB

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -1,29 +1,46 @@
...
             throws RegistrationException {
-        if (repository.findByEmail(request.getEmail()).isPresent()) {
+        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
             throw new RegistrationException("Unable to complete registration.");
         }
         User user = mapper.toModel(request);
-        User savedUser = repository.save(user);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        user.setRoles(Set.of(getUserRole(Role.RoleName.USER)));
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]: these lines have to be done in mapStruct

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository repository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return repository.findByEmail(email).orElseThrow(
+                () -> new RuntimeException("Can't find user by email" + email));
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]:
```suggestion
                () -> new RuntimeException("Can't find user by email: " + email));
```

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml (hunk 3 above)
Comment (koval-kn) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml:22]:
```suggestion
                    references: roles(id)
          primaryKey:
              columnNames: user_id, role_id
```

**Review by koval-kn — CHANGES_REQUESTED** [review]

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**